[#25] 로그인 기능

hyozkim opened the pull request:
- 기능
    - 기본 로그인
    - 네이버 로그인
    - 카카오 로그인
- 테스트
    - 로그인 후 loginCount +1, lastLoginAt 날짜/시간 갱신
    - Api 호출이 필요해서 Mock, Service에서 테스트가 불가능 -> 방법이 있는지 잘 모르겠어요..

hyozkim committed:
기본 로그인, 네이버 로그인 기능 개발

hyozkim committed:
update gradle - jsonObject 라이브러리 추가

hyozkim committed:
Password -> JsonIgnore

hyozkim committed:
code convention

hyozkim committed:
refactoring: naver login logic
- create @Bean: NaverLoginApi, NaverProfileApi

hyozkim committed:
add function: kakao login logic
- create @Bean: KakaoLoginApi, KakaoProfileApi
- write bussiness kakao login logic

ike5923 commented on back-end/src/main/java/com/bangstagram/user/controller/OAuthRestController.java:
```diff
@@ -0,0 +1,47 @@
+     *
+     *  클라이언트에서 요청 -> redirectURI 주소인 oauth/naver로 code, state 받아옴
+     */
+    @GetMapping("oauth/naver")
```
url은 절대주소로

ike5923 commented on back-end/src/main/java/com/bangstagram/user/controller/OAuthRestController.java:
```diff
@@ -0,0 +1,47 @@
+     */
+    @GetMapping("oauth/naver")
+    public AuthResponseDto authNaver(@RequestParam("code") String code,
+                                @RequestParam("state") String state) throws ParseException {
```
exception handling

ike5923 commented on back-end/src/main/java/com/bangstagram/user/domain/model/api/request/JoinRequestDto.java:
```diff
@@ -28,7 +28,7 @@ public JoinRequestDto(String name, String loginEmail, String loginPassword) {
 
-    public User newUser(PasswordEncoder passwordEncoder) {
-        return new User(name,loginEmail,passwordEncoder.encode(loginPassword));
+    public User newUser(PasswordEncoder passwordEncoder, String oAuth) {
```
static

ike5923 commented on back-end/src/main/java/com/bangstagram/user/domain/model/oauth/kakao/KakaoLoginApi.java:
```diff
@@ -0,0 +1,80 @@
+        return new StringBuilder(url)
+                .append("?grant_type="+"authorization_code")
+                .append("&client_id="+clientId)
+                .append("&code="+code).toString();
```
toString은 아래로 내리기

BaeJi77 commented on back-end/src/main/java/com/bangstagram/user/domain/model/user/User.java:
```diff
@@ -31,24 +35,39 @@
+            throw new IllegalArgumentException("Bad Creditials");
+    }
+
+    public void afterLoginSuccess() {
+        this.loginCount++;
+        lastLoginAt = now();
+    }
```
loginCount 이 동작은 왜 하는 것인지 궁금합니다.

BaeJi77 commented on back-end/src/main/java/com/bangstagram/user/util/OAuthUtils.java:
```diff
@@ -0,0 +1,87 @@
+
+public class OAuthUtils {
+
+    public static String get(String apiUrl, Map<String, String> requestHeaders, String contentType){
```
이름이 조금 명확하지 않는 것 같아요. 어떤 것을 get 하는지 알게되면 좋을 것 같아요

BaeJi77 commented on back-end/src/main/java/com/bangstagram/user/domain/model/oauth/kakao/KakaoProfileApi.java:
```diff
@@ -0,0 +1,39 @@
+        this.url = url;
+    }
+
+    public String profileApiUrl() {
+        return url;
+    }
```
이거는 get 쓰시지 않고 이것을 쓴 이유가 궁금합니다.

ike5923 commented on back-end/src/main/java/com/bangstagram/user/service/OAuthService.java:
```diff
@@ -0,0 +1,98 @@
+
+        //TODO: (String) naverProfileApiResult -> JSONObject로 변경, email가져오기
+        NaverProfileApi.UserInfo userInfo = new NaverProfileApi.UserInfo(naverProfileApiResult);
+        String name = (String) userInfo.getName();
```
String.valueOf

BaeJi77 commented on back-end/src/main/java/com/bangstagram/user/controller/OAuthRestController.java:
```diff
@@ -0,0 +1,47 @@
+     *  클라이언트에서 요청 -> redirectURI 주소인 oauth/naver로 code, state 받아옴
+     */
+    @GetMapping("oauth/naver")
+    public AuthResponseDto authNaver(@RequestParam("code") String code,
+                                @RequestParam("state") String state) throws ParseException {
+        log.info("code:{}, state: {}", code, state);
```
여기서 code, state가 값이 없으면 에러가 나는 건가요?
여기서 ParseException을 내는 이유가 뭔가요?

ike5923 commented on back-end/src/main/java/com/bangstagram/user/service/OAuthService.java:
```diff
@@ -0,0 +1,98 @@
+        this.kakaoProfileApi = kakaoProfileApi;
+    }
+
+    public AuthResponseDto loginWithNaver(String code, String state) throws ParseException {
```
메소드 추출해서 읽기좋게

ike5923 commented on back-end/src/main/java/com/bangstagram/user/service/UserService.java:
```diff
@@ -56,5 +57,26 @@ public boolean existsByEmail(String email) {
 
+    @Transactional
+    public AuthResponseDto login(String email, String password) {
+        User user = findByEmail(email).orElseThrow(() -> new RuntimeException("email not found"));
```
exception을 custom하게

ike5923 replied:
공통되는 메소드는 추출해서 사용

BaeJi77 commented on back-end/src/main/resources/application.yml:
```diff
@@ -47,4 +47,15 @@ jwt:
\ No newline at end of file
+    expirySeconds: 0
+
+oauth:
+  naver:
+    clientId: XtF1W2Q5emLFga3N4wv0
+    clientSecret: GDb74_7kCw
+    tokenRequestUrl: https://nid.naver.com/oauth2.0/token
+    profileRequestUrl: https://openapi.naver.com/v1/nid/me
+  kakao:
+    clientId: 97baa0b614b0cff3ba0c550e515bfffb
+    tokenRequestUrl: https://kauth.kakao.com/oauth/token
+    profileRequestUrl: https://kapi.kakao.com/v2/user/me
```
이거 원래 이렇게 공개해도 괜찮은가요?

BaeJi77 commented on back-end/build.gradle:
```diff
@@ -28,6 +28,7 @@ dependencies {
     implementation 'org.apache.commons:commons-lang3:3.10'
     implementation 'org.seleniumhq.selenium:selenium-java:3.141.59'
     implementation 'org.projectlombok:lombok'
+    compile group: 'com.googlecode.json-simple', name: 'json-simple', version: '1.1'
```
이거 너무 옛날 버전이라서 org.json 이거 사용하는게 좋을 것 같아요!

ike5923 commented on back-end/src/main/java/com/bangstagram/user/util/OAuthUtils.java:
```diff
@@ -0,0 +1,87 @@
+            int responseCode = con.getResponseCode();
+            if (responseCode == HttpURLConnection.HTTP_OK) { // 정상 호출
+                return readBody(con.getInputStream());
+            } else { // 에러 발생
```
else문은 사용하지 않는게 좋다

hyozkim committed:
코드 리뷰 후 소스 수정
- 컨벤션(intellij 자동정렬), 네이밍 수정(get~)
- org.json.JSONObject 클래스로 변경
- Dto 에서 Validation
- throws Conroller까지 올라오지 말고, 그곳에서 try catch 쓰기.(어디서 에러 났는지 구체적이지 못하다)
- now() -> static 메소드 클래스 앞에 붙여서 사용. (import static X)
- Optional 사용

hyozkim committed:
Controller 에서 Dto 받기

hyozkim committed:
update 의 경우, save 할 필요 없음. (jpa 종속성 때문에 됨)

hyozkim committed:
url 절대주소로

hyozkim committed:
Merge branch 'develop' into feature/25_signin
